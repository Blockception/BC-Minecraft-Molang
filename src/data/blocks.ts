import { Data } from "./data";

/**Molang data for blocks*/
export namespace Blocks {
  /**The list of usable variables for this specific type*/
  export const Variables: Data[] = [];
  /**The list of specific contexts usable for this specific type*/
  export const Contexts: Data[] = [];
    /**The list of usable temp variables for this specific type*/
  export const Temps: Data[] = [];
}
