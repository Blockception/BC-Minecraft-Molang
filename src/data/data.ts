import { Types } from "bc-minecraft-bedrock-types";

/**The base type of a molang data item*/
export type Data = Types.Identifiable & Types.Documentated;
