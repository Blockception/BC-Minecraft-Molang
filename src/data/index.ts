/*	Auto generated	*/
export * from "./animations";
export * from "./animations-controllers";
export * from "./attachables";
export * from "./blocks";
export * from "./data";
export * from "./entities";
export * from "./features-rules";
export * from "./general";
export * from "./items";
export * from "./molang-data";
export * from "./molang-function";
export * from "./particles";
export * from "./render-controllers";

