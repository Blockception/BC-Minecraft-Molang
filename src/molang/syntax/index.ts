export * from './builder'
export * from './errors'
export * from './nodes'
export * from './operators'
export * from './parse'
export * from './tokens'
export * from './util'