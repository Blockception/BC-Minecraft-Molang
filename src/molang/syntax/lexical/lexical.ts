import { OffsetWord } from "bc-minecraft-bedrock-types/lib/types";
import { isNumberOrLetter } from './util';

export class ParseError extends Error {
  offset: number;
  index: number;
  base: Partial<Error>;

  constructor(message: string, base: Partial<Error>, index: number, offset: number) {
    super(message);
    this.base = base;
    this.name = "molang-parse-error";
    this.stack = base.stack;
    this.index = index;
    this.offset = offset;
  }
}

export function parseLexical(line: OffsetWord | string): LexicalNode[] {
  if (typeof line === "string") {
    const parser = new Parser(line);
    try {
      parser.parse();
    } catch (err: any) {
      throw new ParseError(err.message ?? "parsing error occured", err, parser.index, 0);
    }

    return parser.result;
  }

  const parser = new Parser(line.text);

  try {
    parser.parse();
    // Process offset
    parser.moveOffset(line.offset);
  } catch (err: any) {
    throw new ParseError(err.message ?? "parsing error occured", err, parser.index, 0);
  }

  return parser.result;
}

export enum Token {
  unknown,
  access,
  assignment,
  constant,
  identifier,
  keyword,
  operator,
  punction,
  number,
  text,
  end,
}

export interface LexicalNode {
  text: string;
  offset: number;
  type: Token;
}

class Parser {
  result: LexicalNode[];
  start: number;
  index: number;
  text: string;
  offset: number;

  constructor(text: string) {
    this.result = [];
    this.text = text;
    this.start = 0;
    this.index = 0;
    this.offset = 0;
  }

  push(text: string, offset: number, type: Token): LexicalNode {
    const item = {
      text,
      offset: this.offset + offset,
      type,
    };
    this.result.push(item);
    return item;
  }

  moveIndex(amount: number) {
    this.index += amount;
  }

  moveStart(amount: number) {
    this.start += amount;
  }

  setStart(index: number) {
    this.start = index;
  }

  /**
   * Adds an offset to each item in the parser present
   * @param amount
   */
  moveOffset(amount: number) {
    this.result.forEach((item) => (item.offset += amount));
  }

  /** Assumes that the character at index needs to be isolated into its own node */
  isolateCurrent(length: number, charType: Token) {
    const endIndex = this.index + length;
    // We know that all text before this is its own thing, the found character its own thing and the rest still needs to be parsed
    this.push(this.text.slice(0, this.index), this.start, Token.unknown);
    this.push(this.text.slice(this.index, endIndex), this.index, charType);

    // Move our location further ahead, and put the indexes back at the beginnning;
    this.offset += endIndex;
    this.text = this.text.slice(endIndex);
    this.index = 0;
    this.start = 0;
  }

  parse() {
    // Moves along the text, until we can find things that we can split, then we identify and move further
    while (this.text.length > 0) {
      if (isNumberOrLetter(this.text.charCodeAt(this.index))) {
        this.index++;
        continue;
      }

      this.parseAt();
    }
  }

  // Does one round of parsing
  private parseAt() {
    let c = this.text.slice(this.index, this.index + 2);
    switch (c) {
      case "": // End of the string
      case undefined:
        this.push(this.text, this.start, Token.unknown);
        this.text = "";
        return;

      case "==":
      case "!":
      case "||":
      case "&&":
      case "<":
      case "<=":
      case ">=":
      case ">":
      case "!=":
      case "??":
        return this.isolateCurrent(2, Token.operator);
      case "->":
        return this.isolateCurrent(length, Token.access);
    }

    // Character that we can cut around, like operators
    c = c.slice(0, 1);
    switch (c) {
      case "": // End of the string
      case undefined:
        this.push(this.text, this.start, Token.unknown);
        this.text = "";
        return;

      // Basic math operators
      case "-":
      case "!":
      case "?":
      case "*":
      case "/":
      case "+":
      case "<":
      case ">":
      case ":":
        return this.isolateCurrent(1, Token.operator);
      case ";":
      case ",":
      case ".":
        return this.isolateCurrent(1, Token.punction);
      case "'":
      case '"':
      case "(":
      case ")":
      case "[":
      case "]":
      case "{":
      case "}":
        return this.isolateCurrent(1, Token.unknown);
      case "=":
        return this.isolateCurrent(1, Token.assignment);
    }

    this.moveIndex(1);
  }
}
