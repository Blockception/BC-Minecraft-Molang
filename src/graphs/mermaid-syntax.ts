import { Node, Syntax } from "../molang/syntax/node";

// stateDiagram-v2
//         [*] --> Still
//         Still --> [*]
//         Still --> Moving
//         Moving --> Still
//         Moving --> Crash
//         Crash --> [*]

export function syntaxToStageDiagram(code: Syntax): string {
  let gid = 0;
  function getId() {
    return gid++;
  }

  const lines: string[] = [`[*] --> t0`];

  function syntaxToState(code: Syntax): number {
    const rootId = getId();
    switch (code.type) {
      case Node.function:
        lines.push(bType(rootId, code));
        const identifier = syntaxToState(code.identifier);
        const fparameters = code.parameters.map(syntaxToState);
        lines.push(link(rootId, identifier));
        fparameters.forEach((i) => lines.push(link(rootId, i)));

        break;
      case Node.access:
        lines.push(item(rootId, code.accessor.text), bType(rootId, code));
        const base = syntaxToState(code.base);
        const property = syntaxToState(code.property);
        lines.push(link(rootId, base), link(base, property));
        break;
      case Node.assignment:
        lines.push(item(rootId, "="), bType(rootId, code));
        const receiver = syntaxToState(code.receiver);
        const value = syntaxToState(code.value);
        lines.push(link(value, rootId), link(rootId, receiver));
        break;
      case Node.compare:
        lines.push(item(rootId, code.operator.text), bType(rootId, code));
        const a = syntaxToState(code.a);
        const b = syntaxToState(code.b);
        lines.push(link(rootId, a), link(rootId, b));
        break;
      case Node.identifier:
      case Node.unknown:
      case Node.constant:
        lines.push(item(rootId, code.base.text), bType(rootId, code));
        break;
      case Node.operation:
        lines.push(item(rootId, code.operation.text), bType(rootId, code));
        const parameters = code.parameters.map(syntaxToState);
        parameters.forEach((i) => lines.push(link(rootId, i)));
        break;
    }

    return rootId;
  }

  syntaxToState(code);

  return `stateDiagram-v2\n${lines
    .filter((item) => !item.endsWith(": unknown"))
    .map((item) => `    ${item}`)
    .join("\n")}`;
}

function item(id: number, text: string) {
  return `t${id}: ${text}` as const;
}
function bType(id: number, code: Syntax) {
  return `t${id}: ${Node[code.type]}` as const;
}

function link(fromId: number, toId: number) {
  return `t${fromId} --> t${toId}` as const;
}
