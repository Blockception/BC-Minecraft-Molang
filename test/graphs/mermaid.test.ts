import { OffsetWord } from "bc-minecraft-bedrock-types/lib/types";
import { toBlocks } from "../../src/molang/syntax/lexical/blocks";
import { parse } from "../../src/molang/syntax/lexical/parsing";
import { blocksToMermaid } from "../../src/graphs/mermaid-blocks";
import { valid_syntaxes } from "../data/dataset";
import { parseMolang } from "../../src/molang/syntax/parse";
import { syntaxToStageDiagram } from "../../src/graphs/mermaid-syntax";

describe("mermiad block graphs", () => {
  describe("should be able generate a mermaid flowchart of lexical blocks", () => {
    test.each(valid_syntaxes)("%#. %s", (s) => {
      const tokens = parse(OffsetWord.create(s, 0));
      const blocks = toBlocks(tokens);

      blocks.forEach((item) => {
        expect(blocksToMermaid(item)).toMatchSnapshot();
      });
    });
  });
});

describe("mermiad state diagrams", () => {
  describe("should be able generate a mermaid stategiram of syntax", () => {
    test.each(valid_syntaxes)("%#. %s", (s) => {
      const statements = parseMolang(OffsetWord.create(s));

      statements.forEach((item) => {
        expect(syntaxToStageDiagram(item)).toMatchSnapshot();
      });
    });
  });
});
