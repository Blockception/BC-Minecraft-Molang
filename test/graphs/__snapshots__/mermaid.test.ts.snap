// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 0. variable.agent.anim_state == 2 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.agent.anim_state"]
    T2["=="]
    T3["\\2"]
    end

    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 1. variable.agent.anim_state > 2 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.agent.anim_state"]
    T2["\\>"]
    T3["\\2"]
    end

    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 2. variable.agent.anim_state < 2 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.agent.anim_state"]
    T2["\\<"]
    T3["\\2"]
    end

    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 3. variable.agent.anim_state != 2 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.agent.anim_state"]
    T2["!="]
    T3["\\2"]
    end

    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 4. variable.agent.anim_state >= 2 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.agent.anim_state"]
    T2[">="]
    T3["\\2"]
    end

    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 5. variable.agent.anim_state <= 2 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.agent.anim_state"]
    T2["<="]
    T3["\\2"]
    end

    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 6. !variable.swaying && query.all_animations_finished 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["\\!"]
    subgraph T2["block of ()"]
    direction LR
    T3["variable.swaying"]
    T4["&&"]
    T5["query.all_animations_finished"]
    end
    end

    T3 --> T4
    T4 --> T5
    T1 --> T2"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 7. 1.5 - math.pow(1.5, 1.0 - 015.0 * query.modified_move_speed) * query.camera_rotation(0) 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["1.5"]
    T2["\\-"]
    subgraph T3["block of ()"]
    direction LR
    subgraph T4["block of ()"]
    direction LR
    T5["math.pow"]
    subgraph T6["block of ()"]
    direction LR
    T7["1.5"]
    T8["1.0-015.0*query.modified_move_speed"]
    end
    end
    T9["\\*"]
    subgraph T10["block of ()"]
    direction LR
    T11["query.camera_rotation"]
    subgraph T12["block of ()"]
    direction LR
    T13["\\0"]
    end
    end
    end
    end

    T7 --> T8
    T5 --> T6
    T11 --> T12
    T4 --> T9
    T9 --> T10
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 8. math.clamp(math.sqrt(math.pow(query.position_delta(0), 2.0) + math.pow(query.position_delta(2), 2.0)), 0.0, 0.1) 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["math.clamp"]
    subgraph T2["block of ()"]
    direction LR
    subgraph T3["block of "]
    direction LR
    T4["math.sqrt"]
    subgraph T5["block of ()"]
    direction LR
    subgraph T6["block of "]
    direction LR
    T7["math.pow"]
    subgraph T8["block of ()"]
    direction LR
    subgraph T9["block of "]
    direction LR
    T10["query.position_delta"]
    subgraph T11["block of ()"]
    direction LR
    T12["\\0"]
    end
    T13[",2.0"]
    end
    end
    T14["+math.pow"]
    subgraph T15["block of ()"]
    direction LR
    subgraph T16["block of "]
    direction LR
    T17["query.position_delta"]
    subgraph T18["block of ()"]
    direction LR
    T19["\\2"]
    end
    T20[",2.0"]
    end
    end
    end
    end
    T21[",0.0,0.1"]
    end
    end
    end

    T10 --> T11
    T11 --> T13
    T17 --> T18
    T18 --> T20
    T7 --> T8
    T8 --> T14
    T14 --> T15
    T4 --> T5
    T5 --> T21
    T1 --> T2"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 9. -175 * math.sin(variable.attack_time * 180) 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["\\-"]
    subgraph T2["block of ()"]
    direction LR
    T3["175"]
    T4["\\*"]
    subgraph T5["block of ()"]
    direction LR
    T6["math.sin"]
    subgraph T7["block of ()"]
    direction LR
    T8["variable.attack_time*180"]
    end
    end
    end
    end

    T6 --> T7
    T3 --> T4
    T4 --> T5
    T1 --> T2"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 10. variable.is_enchanted ? Material.enchanted : Material.default 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    subgraph T1["block of ()"]
    direction LR
    T2["variable.is_enchanted"]
    T3["\\?"]
    T4["Material.enchanted"]
    end
    T5["\\:"]
    T6["Material.default"]
    end

    T2 --> T3
    T3 --> T4
    T1 --> T5
    T5 --> T6"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 11. variable.has_trim ? variable.trim_path : Texture.default 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    subgraph T1["block of ()"]
    direction LR
    T2["variable.has_trim"]
    T3["\\?"]
    T4["variable.trim_path"]
    end
    T5["\\:"]
    T6["Texture.default"]
    end

    T2 --> T3
    T3 --> T4
    T1 --> T5
    T5 --> T6"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 12. !variable.has_trim 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["\\!"]
    T2["variable.has_trim"]
    end

    T1 --> T2"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 13. array.skins[query.property('minecraft:has_nectar') + query.is_angry * 2] 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["array.skins"]
    subgraph T2["block of []"]
    direction LR
    subgraph T3["block of "]
    direction LR
    T4["query.property"]
    subgraph T5["block of ()"]
    direction LR
    T6["'minecraft:has_nectar'"]
    end
    T7["+query.is_angry*2"]
    end
    end
    end

    T4 --> T5
    T5 --> T7
    T1 --> T2"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 14. variable.x_head_rotation = math.clamp(query.target_x_rotation, -32.5, 32.5); 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.x_head_rotation"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["math.clamp"]
    subgraph T5["block of ()"]
    direction LR
    T6["query.target_x_rotation"]
    T7["-32.5"]
    T8["32.5"]
    end
    end
    end

    T6 --> T7
    T7 --> T8
    T4 --> T5
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 15. variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.y_head_rotation"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["math.clamp"]
    subgraph T5["block of ()"]
    direction LR
    T6["query.target_y_rotation"]
    T7["-22.5"]
    T8["25"]
    end
    end
    end

    T6 --> T7
    T7 --> T8
    T4 --> T5
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 16. variable.is_rolled_up = query.property('minecraft:armadillo_state') != 'unrolled'; 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.is_rolled_up"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["query.property"]
    subgraph T5["block of ()"]
    direction LR
    T6["'minecraft:armadillo_state'"]
    end
    T7["!='unrolled'"]
    end
    end

    T4 --> T5
    T5 --> T7
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 17. variable.rolled_up_time = variable.is_rolled_up ? ((variable.rolled_up_time ?? 0.0) + query.delta_time) : 0.0; 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.rolled_up_time"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["variable.is_rolled_up"]
    T5["\\?"]
    subgraph T6["block of ()"]
    direction LR
    subgraph T7["block of ()"]
    direction LR
    T8["variable.rolled_up_time??0.0"]
    T9["+query.delta_time"]
    end
    T10[":0.0"]
    end
    end
    end

    T8 --> T9
    T7 --> T10
    T4 --> T5
    T5 --> T6
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 18. variable.is_peeking = query.property('minecraft:armadillo_state') == 'rolled_up_peeking'; 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.is_peeking"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["query.property"]
    subgraph T5["block of ()"]
    direction LR
    T6["'minecraft:armadillo_state'"]
    end
    T7["=='rolled_up_peeking'"]
    end
    end

    T4 --> T5
    T5 --> T7
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 19. variable.is_unrolling = query.property('minecraft:armadillo_state') == 'rolled_up_unrolling'; 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.is_unrolling"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["query.property"]
    subgraph T5["block of ()"]
    direction LR
    T6["'minecraft:armadillo_state'"]
    end
    T7["=='rolled_up_unrolling'"]
    end
    end

    T4 --> T5
    T5 --> T7
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 20. variable.unrolling_time = variable.is_unrolling ? ((variable.unrolling_time ?? 0.0) + query.delta_time) : 0.0; 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.unrolling_time"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["variable.is_unrolling"]
    T5["\\?"]
    subgraph T6["block of ()"]
    direction LR
    subgraph T7["block of ()"]
    direction LR
    T8["variable.unrolling_time??0.0"]
    T9["+query.delta_time"]
    end
    T10[":0.0"]
    end
    end
    end

    T8 --> T9
    T7 --> T10
    T4 --> T5
    T5 --> T6
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 21. variable.use_rolled_up_model = variable.rolled_up_time >= 0.2083 && (variable.unrolling_time == 0.0 || variable.unrolling_time <= 1.25 ); 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.use_rolled_up_model"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    subgraph T4["block of ()"]
    direction LR
    T5["variable.rolled_up_time"]
    T6[">="]
    T7["0.2083"]
    end
    T8["&&"]
    subgraph T9["block of ()"]
    direction LR
    T10["variable.unrolling_time==0.0||variable.unrolling_time<=1.25"]
    end
    end
    end

    T5 --> T6
    T6 --> T7
    T4 --> T8
    T8 --> T9
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 22. variable.walking = query.modified_move_speed > 0.01 && !variable.is_rolled_up; 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.walking"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    subgraph T4["block of ()"]
    direction LR
    T5["query.modified_move_speed"]
    T6["\\>"]
    T7["0.01"]
    end
    T8["&&"]
    subgraph T9["block of ()"]
    direction LR
    T10["\\!"]
    T11["variable.is_rolled_up"]
    end
    end
    end

    T5 --> T6
    T6 --> T7
    T10 --> T11
    T4 --> T8
    T8 --> T9
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 23. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time; 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.walk_anim_time_update"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["query.anim_time"]
    T5["\\+"]
    subgraph T6["block of ()"]
    direction LR
    T7["math.min"]
    subgraph T8["block of ()"]
    direction LR
    T9["3.0"]
    subgraph T10["block of "]
    direction LR
    T11["math.lerp"]
    subgraph T12["block of ()"]
    direction LR
    T13["2.0"]
    T14["5.0"]
    T15["query.modified_move_speed"]
    end
    end
    end
    T16["*query.delta_time"]
    end
    end
    end

    T13 --> T14
    T14 --> T15
    T11 --> T12
    T9 --> T10
    T7 --> T8
    T8 --> T16
    T4 --> T5
    T5 --> T6
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 24. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time;variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 1`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.walk_anim_time_update"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["query.anim_time"]
    T5["\\+"]
    subgraph T6["block of ()"]
    direction LR
    T7["math.min"]
    subgraph T8["block of ()"]
    direction LR
    T9["3.0"]
    subgraph T10["block of "]
    direction LR
    T11["math.lerp"]
    subgraph T12["block of ()"]
    direction LR
    T13["2.0"]
    T14["5.0"]
    T15["query.modified_move_speed"]
    end
    end
    end
    T16["*query.delta_time"]
    end
    end
    end

    T13 --> T14
    T14 --> T15
    T11 --> T12
    T9 --> T10
    T7 --> T8
    T8 --> T16
    T4 --> T5
    T5 --> T6
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad block graphs should be able generate a mermaid flowchart of lexical blocks 24. variable.walk_anim_time_update = query.anim_time + math.min(3.0, math.lerp(2.0, 5.0, query.modified_move_speed)) * query.delta_time;variable.y_head_rotation = math.clamp(query.target_y_rotation, -22.5, 25); 2`] = `
"flowchart LR
    subgraph T0["block of "]
    direction LR
    T1["variable.y_head_rotation"]
    T2["\\="]
    subgraph T3["block of ()"]
    direction LR
    T4["math.clamp"]
    subgraph T5["block of ()"]
    direction LR
    T6["query.target_y_rotation"]
    T7["-22.5"]
    T8["25"]
    end
    end
    end

    T6 --> T7
    T7 --> T8
    T4 --> T5
    T1 --> T2
    T2 --> T3"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 0. variable.agent.anim_state == 2 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: ==
    t0: compare
    t1: .
    t1: access
    t2: variable
    t3: .
    t3: access
    t4: agent
    t5: anim_state
    t3 --> t4
    t4 --> t5
    t1 --> t2
    t2 --> t3
    t6: 2
    t6: constant
    t0 --> t1
    t0 --> t6"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 1. variable.agent.anim_state > 2 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: >
    t0: compare
    t1: .
    t1: access
    t2: variable
    t3: .
    t3: access
    t4: agent
    t5: anim_state
    t3 --> t4
    t4 --> t5
    t1 --> t2
    t2 --> t3
    t6: 2
    t6: constant
    t0 --> t1
    t0 --> t6"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 2. variable.agent.anim_state < 2 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: <
    t0: compare
    t1: .
    t1: access
    t2: variable
    t3: .
    t3: access
    t4: agent
    t5: anim_state
    t3 --> t4
    t4 --> t5
    t1 --> t2
    t2 --> t3
    t6: 2
    t6: constant
    t0 --> t1
    t0 --> t6"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 3. variable.agent.anim_state != 2 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: !=
    t0: compare
    t1: .
    t1: access
    t2: variable
    t3: .
    t3: access
    t4: agent
    t5: anim_state
    t3 --> t4
    t4 --> t5
    t1 --> t2
    t2 --> t3
    t6: 2
    t6: constant
    t0 --> t1
    t0 --> t6"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 4. variable.agent.anim_state >= 2 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: >=
    t0: compare
    t1: .
    t1: access
    t2: variable
    t3: .
    t3: access
    t4: agent
    t5: anim_state
    t3 --> t4
    t4 --> t5
    t1 --> t2
    t2 --> t3
    t6: 2
    t6: constant
    t0 --> t1
    t0 --> t6"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 5. variable.agent.anim_state <= 2 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: <=
    t0: compare
    t1: .
    t1: access
    t2: variable
    t3: .
    t3: access
    t4: agent
    t5: anim_state
    t3 --> t4
    t4 --> t5
    t1 --> t2
    t2 --> t3
    t6: 2
    t6: constant
    t0 --> t1
    t0 --> t6"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 6. !variable.swaying && query.all_animations_finished 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: !
    t0: operation
    t1: &&
    t1: operation
    t2: .
    t2: access
    t3: variable
    t4: .
    t4: access
    t5: variable
    t6: swaying
    t4 --> t5
    t5 --> t6
    t2 --> t3
    t3 --> t4
    t7: .
    t7: access
    t8: query
    t9: .
    t9: access
    t10: query
    t11: all_animations_finished
    t9 --> t10
    t10 --> t11
    t7 --> t8
    t8 --> t9
    t1 --> t2
    t1 --> t7
    t0 --> t1"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 10. variable.is_enchanted ? Material.enchanted : Material.default 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: :
    t0: operation
    t1: ?
    t1: compare
    t2: .
    t2: access
    t3: variable
    t4: .
    t4: access
    t5: variable
    t6: is_enchanted
    t4 --> t5
    t5 --> t6
    t2 --> t3
    t3 --> t4
    t7: .
    t7: access
    t8: Material
    t9: .
    t9: access
    t10: Material
    t11: enchanted
    t9 --> t10
    t10 --> t11
    t7 --> t8
    t8 --> t9
    t1 --> t2
    t1 --> t7
    t12: .
    t12: access
    t13: Material
    t14: .
    t14: access
    t15: Material
    t16: default
    t14 --> t15
    t15 --> t16
    t12 --> t13
    t13 --> t14
    t0 --> t1
    t0 --> t12"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 11. variable.has_trim ? variable.trim_path : Texture.default 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: :
    t0: operation
    t1: ?
    t1: compare
    t2: .
    t2: access
    t3: variable
    t4: .
    t4: access
    t5: variable
    t6: has_trim
    t4 --> t5
    t5 --> t6
    t2 --> t3
    t3 --> t4
    t7: .
    t7: access
    t8: variable
    t9: .
    t9: access
    t10: variable
    t11: trim_path
    t9 --> t10
    t10 --> t11
    t7 --> t8
    t8 --> t9
    t1 --> t2
    t1 --> t7
    t12: .
    t12: access
    t13: Texture
    t14: .
    t14: access
    t15: Texture
    t16: default
    t14 --> t15
    t15 --> t16
    t12 --> t13
    t13 --> t14
    t0 --> t1
    t0 --> t12"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 12. !variable.has_trim 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: !
    t0: operation
    t1: .
    t1: access
    t2: variable
    t3: .
    t3: access
    t4: variable
    t5: has_trim
    t3 --> t4
    t4 --> t5
    t1 --> t2
    t2 --> t3
    t0 --> t1"
`;

exports[`mermiad state diagrams should be able generate a mermaid stategiram of syntax 22. variable.walking = query.modified_move_speed > 0.01 && !variable.is_rolled_up; 1`] = `
"stateDiagram-v2
    [*] --> t0
    t0: =
    t0: assignment
    t1: .
    t1: access
    t2: variable
    t3: .
    t3: access
    t4: variable
    t5: walking
    t3 --> t4
    t4 --> t5
    t1 --> t2
    t2 --> t3
    t6: &&
    t6: operation
    t7: >
    t7: compare
    t8: .
    t8: access
    t9: query
    t10: .
    t10: access
    t11: query
    t12: modified_move_speed
    t10 --> t11
    t11 --> t12
    t8 --> t9
    t9 --> t10
    t13: 0.01
    t13: constant
    t7 --> t8
    t7 --> t13
    t14: !
    t14: operation
    t15: .
    t15: access
    t16: variable
    t17: .
    t17: access
    t18: variable
    t19: is_rolled_up
    t17 --> t18
    t18 --> t19
    t15 --> t16
    t16 --> t17
    t14 --> t15
    t6 --> t7
    t6 --> t14
    t6 --> t0
    t0 --> t1"
`;
